File: android/src/main/java/com/lightningkite/kwift/observables/binding/Button.binding.actual.kt
package com.lightningkite.kwift.observables.binding

import android.graphics.drawable.Drawable
import android.widget.Button
import com.lightningkite.kwift.observables.ObservableProperty
import com.lightningkite.kwift.observables.addAndRunWeak
import com.lightningkite.kwift.views.ColorResource
import com.lightningkite.kwift.observables.*
import com.lightningkite.kwift.views.backgroundDrawable


fun Button.bindActive(
    observable: ObservableProperty<Boolean>,
    activeColorResource: ColorResource? = null,
    inactiveColorResource: ColorResource? = null
) {
    observable.addAndRunWeak(this) { self, it ->
        self.isEnabled = it
        if (it) {
            activeColorResource?.let { color ->
                self.setBackgroundResource(color)
            }
        } else {
            inactiveColorResource?.let { color ->
                self.setBackgroundResource(color)
            }
        }
    }
}

fun Button.bindActive(
    observable: ObservableProperty<Boolean>,
    activeBackground: Drawable,
    inactiveBackground: Drawable
) {
    observable.addAndRunWeak(this) { self, it ->
        self.isEnabled = it
        if (it) {
            self.backgroundDrawable = activeBackground
        } else {

            self.backgroundDrawable = inactiveBackground
        }
    }
}

File: ios/Kwift/com/lightningkite/kwift/observables/binding/Button.binding.actual.swift
//Stub file made with Kwift 2 (by Lightning Kite)
import Foundation
import UIKit


//--- Button.bindActive(ObservableProperty<Boolean>, ColorResource? , ColorResource? )
public extension UIButton {
    func bindActive(_ observable: ObservableProperty<Bool>, _ activeColorResource: ColorResource? = nil, _ inactiveColorResource: ColorResource? = nil) -> Void {
        observable.addAndRunWeak(referenceA: self) { (this, value) in
            this.isUserInteractionEnabled = value
            if value {
                if let color = activeColorResource {
                    this.backgroundColor = color
                }
            }else{
                if let color = inactiveColorResource{
                    this.backgroundColor = color
                }
            }
        }
    }
    func bindActive(observable: ObservableProperty<Bool>, activeColorResource: ColorResource? = nil, inactiveColorResource: ColorResource? = nil) -> Void {
        return bindActive(observable, activeColorResource, inactiveColorResource)
    }
}





//--- Button.bindActive(ObservableProperty<Boolean>, Drawable, Drawable)
public extension UIButton {
    func bindActive(_ observable: ObservableProperty<Bool>, _ activeBackground: @escaping Drawable, _ inactiveBackground: @escaping Drawable) -> Void {
        observable.addAndRunWeak(referenceA: self) { (this, value) in
            this.isUserInteractionEnabled = value
            if value {
                this.backgroundDrawable = activeBackground
            }else{
                this.backgroundDrawable = inactiveBackground
            }
        }
    }
    func bindActive(observable: ObservableProperty<Bool>, activeBackground: @escaping Drawable, inactiveBackground: @escaping Drawable) -> Void {
        return bindActive(observable, activeBackground, inactiveBackground)
    }
}





